service/dataset.py

```diff
@@ -48,60 +76,151 @@ def read(dataset_request):
     except:
         # TODO: Temp fix for testing - Do better cleanup if secret missing
         # dataset["dataset_type"]["token"] = {'name':dataset_name,'value':42}
-        pass 
+        pass
+
+    return dataset
+
+def read(dataset_request):
+    """
+    Read from private dataset
+
+    :param info: The dataset to read and budget to use.
+    :type info: dict {"dataset_name": str, "budget":int}
+    :return: A dataset document that contains the type and info of the dataset
+    :rtype: dict{"dataset_type": str, dataset_key: dict}
+    """
+    return _read_helper(dataset_request, DATASETS)
+
+def readreleased(dataset_request):
+    """
+    Read from a released dataset, if authorized
     
-    # Check/Decrement the budget before returning dataset
-    adjusted_budget = dataset["budget"] - dataset_request["budget"]
-    if adjusted_budget >= 0.0:
-        dataset["budget"] = adjusted_budget
-    else:
-        abort(412, "Not enough budget for read. Remaining budget: {}".format(dataset_name))
+    Check request header for client_guid
 
+    :param info: The dataset to read and budget to use
+    :type info: dict {"dataset_name": str, "budget":int}
+    :return: A dataset document that contains the type and info of the dataset
+    :rtype: dict{"dataset_type": str, dataset_key: dict}
+    """
+    dataset = _read_helper(dataset_request, RELEASED_DATASETS)
+    dataset_name = dataset["dataset_name"]
+
+    client_guid = request.headers.get('client_guid')
+
+    # If client owns a released dataset, treat readreleased as
+    # private read
+    if client_guid in OWNERS:
+        if dataset_request["dataset_name"] in OWNERS[client_guid]:
+            return dataset
+
+    # Checks to see if authorized users specified
+    if 'authorized_users' in dataset and len(dataset['authorized_users']) > 0:
+        if client_guid not in dataset['authorized_users']:
+            abort(404, "User not authorized to access this dataset.")
+
+    # Track readrelease
+    USERS_RELEASED[dataset_name][client_guid] += 1
+
+    # Decrement budget if this is a users first read
+    if USERS_RELEASED[dataset_name][client_guid] > 1:
```

Comment:
I don't think we need this section until we have per user budget. With global budget the cost of the dataset was used. We can keep user incrementing since it will be important later

Reply:
Also this counts as state so we would like to avoid it, we will probably have to do an accounting document for user reads of released datasets longer term. This is good for our reference implementation

Reply:
removed for now - will have to add user accounting in future PR